### ISSUE TITLE
feature request: "allowEmpty" option for `arrow-body-style`

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#5159]
Basically, I want an option so that:
```js
const foo = () => {};
```

is not an error, but:
```js
const foo = () => { a: 1 };
```
is.

In other words, "an arrow function that might intend to return an empty object" is not a thing I'm worried about - and in fact i *want* to use that as a noop function that returns `undefined` - but "an arrow function that might intend to return a non-empty object literal" is a thing I'm very worried about.

**Comment by eslintbot** [comment]
@ljharb Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ljharb** [comment]
)I'd like this option in both 1.x and 2.x)

**Comment by mysticatea** [comment]
Thank you for this issue.

Actually, now the rule is not warning empty blocks.
Related: #4411

The `const foo = () => { a: 1 };` pattern is warned by other rules such as `no-labels`, `no-unused-labels`, and `no-unused-expression`.
So I'm not sure whether this rule should warn it or not.

**Comment by michaelficarra** [comment]
Agreed with @mysticatea. We already have a number of more general rules that will catch this error.

**Comment by ljharb** [comment]
Sorry if I was unclear - this rule *already* warns on that, but it also warns when I use explicit return when the body is returning a single expression.

I want an option so that I can preserve "warn when I use explicit return when the body is returning a single expression" but skip *only* the common noop pattern `() => {}`.

**Comment by michaelficarra** [comment]
What rule are you talking about?

**Comment by ljharb** [comment]
`arrow-body-style` (in the title)

**Comment by michaelficarra** [comment]
From the rule documentation,

> Additionally, this rule specifically warns against a possible developer error when the intention is to return an empty object literal but creates an empty block instead, returning undefined.

You want an option to disable this additional check? I'd say it's probably best to split it out to its own rule.

**Comment by ljharb** [comment]
That would be fine also - whatever ends up with me being able to use `arrow-body-style` without the additional "empty block" check :-)

**Comment by mysticatea** [comment]
I meant, the additional empty block check had been removed already.

**Comment by ljharb** [comment]
@mysticatea that doesn't appear to be in 1.x. Has it been backported?

**Comment by mysticatea** [comment]
We have not backported changes in 2.00 pre-release yet.

@nzakas do we have a plan?

**Comment by gyandeeps** [comment]
I am guessing [same comment](https://github.com/eslint/eslint/issues/5102#issuecomment-178061768) applies here.

**Comment by mysticatea** [comment]
Ah, sorry, that has 'breaking' label, so we cannot backport that.

**Comment by ljharb** [comment]
Why must it be breaking? Can't it just be opt-in behavior?

**Comment by michaelficarra** [comment]
I believe @nzakas has said in the past that he considers the removal of a warning a breaking change, not just the addition of one. I disagree with this.

**Comment by ljharb** [comment]
I'm not requesting the default behavior change in any way - I'm requesting a new option be added. That qualifies as semver-minor imo.

**Comment by michaelficarra** [comment]
But @ljharb for 2.0 it appears to already have been removed. We wouldn't *also* add an option to disable it.

**Comment by ljharb** [comment]
Ah - I guess I'm asking for the removal to be guarded by an option, in 1.x only. ie, it'd be a minor in 1.x, and dropping the option guard in 2.x is the major bump.

**Comment by michaelficarra** [comment]
I think you should forget about 1.x. As I understand, it's not going to be maintained separately.

**Comment by ljharb** [comment]
It's unfortunate that the possibility of a clear upgrade path vanished the instant 2.x became a possibility :-/ I'd hope that 1.x would be maintained at least to the point where the transition to a *released* 2.x was as minimally breaking as possible.

**Comment by michaelficarra** [comment]
Maintenance doesn't normally include feature requests.

**Comment by ljharb** [comment]
Sure, but a prerelease also doesn't normally kick a major line into maintenance mode :-)

**Comment by nzakas** [comment]
I'm sorry, I just don't have the extra time, energy, or health to donate to maintaining two separate versions for any period of time. I'm doing my absolute best to make upgrades as easy as possible, but when there's just one (chronically ill) person managing releases, there are severe limits to what is feasible.

In other news, as this seems to be addressed on master, I'm closing this issue.

**Comment by ljharb** [comment]
Thanks, hopefully the upgrade to v2 is smooth once it's released!
